Support for Canon CR3 raw files

cytrinox opened the pull request:
This PR adds support for Canon CR3 raw files. 

 * CRX Decompressor routines are ported from libraw (CrxDecompressor)
 * A new basic ISOBMFF parser is added
 * A new extended ISOBMFF parser for Canon specific BMFF boxes is added (as part of the Cr3Decoder)
 * New cameras.xml definitions for Canon EOS R and EOS M50 added

Canon CR3 files are something new compared to the good old CR2 format. CR2 was based on TIFF. CR3 is based on ISO BMFF.
BMFF organizes data into boxes. A description of all related boxes can be found at https://github.com/lclevy/canon_cr3.

To support CR3, rawspeed needs an ISOBMFF parser to iterate and query boxes. This is implemented in IsoMBoxes.cpp.
ISOBMFF can be extended with custom boxes, so for the custom Canon boxes, a separate parser is required. The most interesting boxes are:
 * **CMP1**: Box for all parameters needed for decoding/decompression (like encoding type, CFA, compression...)
 *  **IAD1**: Box for sensor information (size, crop, black areas)
 * **CTMD**: Box for Canon Timed MetaData, contains the CANONCOLORDATA EXIF tag for white balance coeffs (record 8)

Older cameras has multiple raw quality modes: raw, mraw, sraw with reduced pixel size (=smaller files). Cameras with CR3 support don't shrink the sensor data, instead a wavelet based compression is applied. So pixel dimensions are equal for **RAW** and **CRAW** (compact RAW) mode. It makes no sense to define craw as separate mode in cameras.xml as this is directly handled by the decompressor.

But Canon has introduced another mode: 1.6x crop. Because full-frame cameras like R and R5 supports old EF-s lenses, the camera auto-forces a crop-mode with EF-s. For EF and RF lenses, crop-mode can be manually enabled. If crop is enabled, the sensor size, raw pixel data, crop-areas and black-areas are different - compared to full-frame images. This is dynamically handled by the **IAD1** information. So crop and black areas are automatically applied and must not be specified in cameras.xml.

To get white balance coeffs is a litte bit tricky. As for CR2 files, Canon stores this information in the CANONCOLORDATA EXIF tag. But this tag is not embedded in CMT1, CMT2, CMT3 or CMT4. Instead, this tag is located in CTMD in track 4. CTMD contains multiple records with a specific type. Record type 8 contains EXIF data and the CANONCOLORDATA tag. The **wb_offset** hint in cameras.xml points to the wb coeffs inside CANONCOLORDATA. This offsets are taken from exiftool source-code and validated against sample images.

As I have only access to samples for EOS R and EOS M50 yet, these are the first cameras with full support. B/W levels are extracted from large image sets of all ISO values.

So far, I was able to open any CR3 with this code: RAW, CRAW, 1.6-cropped, even dual-pixel images. 

I hope this PR helps in adding long awaited CR3 support :smiley: 

Fix for #121 #193 #192

cytrinox committed:
rawspeed: Canon EOS R support

cytrinox committed:
rawspeed: Canon EOS M50 support

cytrinox committed:
Add ISO BMFF parser

This patch is based on a draft from Roman Lebedev.

cytrinox committed:
Extend helper functions getI16() and getByteSwapped()

cytrinox committed:
Add CrxDecompressor for CR3 raw files

This decompressor is ported from libraw. Some of the file handling logic
has been replaced by rawspeed::Buffer and derived classes.
Most of the code is untouched.

cytrinox committed:
Add Cr3Decoder for CR3 raw files

This decoder contains the logic to parse a CR3 ISOBMFF file.

matt-maguire commented:
So, I was able to open a couple of CR3 files from my Canon M50 using darktable on my intel macbook, with a couple of small issues:
- because I bought it in Japan, it identifies itself as Canon Kiss M, which is a local brandname of the M50 for the Japanese market. By adding a new entry to the cameras.xml file, copied off the M50 entry, then darktable was able to open the file.
- darktable couldn't find the standard color matrix, so I created an entry for the M50 in adobe_coeff.c (it turns out that according to Adobe, the coefficients for M50 should be identical to the M100 coefficients).
- another issue of the darktable side, the build scripts couldn't properly detect on my mac that I'd compiling in ISOBMFF support in my exiv2 library, so that's something to follow up with darktable team.

great job @cytrinox , it appears to be working very well!

cytrinox commented:
@matt-maguire 
>    because I bought it in Japan, it identifies itself as Canon Kiss M, which is a local brandname of the M50 for the Japanese market. By adding a new entry to the cameras.xml file, copied off the M50 entry, then darktable was able to open the file.

Please share a patch with me to add the camera entry.

>    darktable couldn't find the standard color matrix, so I created an entry for the M50 in adobe_coeff.c (it turns out that according to Adobe, the coefficients for M50 should be identical to the M100 coefficients).

There is already a PR for darktable with M50 coeffs from me: https://github.com/darktable-org/darktable/pull/8927
Must there a second entry for the "Kiss M" or is just M50 for your model?

>    another issue of the darktable side, the build scripts couldn't properly detect on my mac that I'd compiling in ISOBMFF support in my exiv2 library, so that's something to follow up with darktable team.

The darktable PR for ISOBMFF detection was written by myself, too :) https://github.com/darktable-org/darktable/pull/8751 Please let me know what must be fixed or open a PR :smile:

cytrinox commented:
I've got a full Canon R5 sample set today, so I will integrate R5 as well.

@LebedevRI I know you want samples at raw.pixls.us. To create a full cameras.xml entry with your *dngmeta.rb* script, for all ISO levels a sample file is required. So I've requested a sample matrix of:
 * Quality mode
   * RAW
   * CRAW
 * ISO (L (50), 100, 125, 160, ... HI1, Hi2 (or whatever ISO the camera can reach) in 1/3 ISO levels
 * Dualpixel (if camera support it)
   * enabled
   * disabled
 * 1.6 Crop mode
   * enabled
   * disabled

All files are named with a good name, like *Canon_EOS_R_ISO_100_dualpixel_off_nocrop.CR3*.
I generate the camera entries with dngmeta.rb after converting each of the sample to dng with Adobe DNG Converter. So I've lot of data (must be ~10 GiB now). Do you want all these samples at raw.pixls.us? Then I need a bulk upload :smile: 
If you just want a sample for each mode but without ISO variations, I can upload them (if not already a sample is online). I keep the full sets on my disk if nobody wants the data.

matt-maguire commented:
Here is a rawspeed cameras.xml patch for the Canon Kiss M camera.

[cameras_xml.txt](https://github.com/darktable-org/rawspeed/files/6464571/cameras_xml.txt)

cytrinox committed:
rawspeed: Canon EOS R5 support

cytrinox committed:
rawspeed: Canon EOS R6 support

matt-maguire commented:
Rather than create a separate entry for Canon EOS KISS M in cameras.xml, I have added an alias under the M50 entry (since the KISS M is just the Japanese marketing name for the M50 in Japan). This allows the files for the KISS M to be opened, and the id used for the alias allows lensfun to find the appropriate camera profile for lens correction.

I created a PR against the canon_cr3 branch in @cytrinox github repo ( cytrinox/rawspeed#1 ), if he accepts that PR onto his canon_cr3 branch then I think the associated commit should be automatically picked up by this PR.

cytrinox commented:
@matt-maguire I've already reworked the cameras.xml for the Kiss M and will push the changes soon.

cytrinox committed:
rawspeed: Canon EOS RP support

cytrinox committed:
rawspeed: Canon EOS 850D support

cytrinox committed:
rawspeed: Canon EOS-1D X Mark III support

cytrinox committed:
Alias for Canon EOS M50: EOS KISS M

cytrinox committed:
CR3 IAD1 sensor info should not be DEBUG_PRIO_WARNING

cytrinox commented:
Base samples are uploaded to RPU for all cameras included in this PR. If more/other samples needed, please post an update on this PR and @johnny-bit and myself organize them.

For 90D and 850D, the PR https://github.com/darktable-org/darktable/pull/8962 includes information for adobe_coeff.c.

tpinfold commented:
Thanks for all your hard work on this issue.

Thanatomanic commented:
Don't hesitate to look for adobe coefficients or black/white levels in our [camconst.json](https://github.com/Beep6581/RawTherapee/blob/dev/rtengine/camconst.json) file from RawTherapee. It's may help fill in some gaps!

lclevy commented:
Please do you have cr3 samples with crop (1.6x with EF-S lens) ? It is to improve my cr3 docs and parser .
Thanks

cytrinox commented:
> Please do you have cr3 samples with crop (1.6x with EF-S lens) ? It is to improve my cr3 docs and parser .
> Thanks

I've crop samples for all cameras+modes (that supports crop) listed in this PR. Can you contact me at matrix/irc?

lclevy commented:
> > Please do you have cr3 samples with crop (1.6x with EF-S lens) ? It is to improve my cr3 docs and parser .
> > Thanks
> 
> I've crop samples for all cameras+modes (that supports crop) listed in this PR. Can you contact me at matrix/irc?

Which address? Server? Or room?

cytrinox commented:
> > > Please do you have cr3 samples with crop (1.6x with EF-S lens) ? It is to improve my cr3 docs and parser .
> > > Thanks
> > 
> > 
> > I've crop samples for all cameras+modes (that supports crop) listed in this PR. Can you contact me at matrix/irc?
> 
> Which address? Server? Or room?

#pixls.us or #darktable on freenode @cytrinox

johnny-bit commented:
@lclevy - both darktable and pixls.us channels are on freenode irc
on martix itself there's #darktable-management:matrix.org which is usually quiet and only for dt management but can be used as a hub for reaching darktable devs on matrix
also on matrix there's #exiv2-chat:matrix.org with the devs of exiv2 :)

also the call for samples is present (with links still viable and current sample status) here:
https://discuss.pixls.us/t/call-for-extended-cr3-samples-all-iso-combinations-needed/24959?u=johnny-bit

AFAIK cytrinox has all the files sent by users.

lclevy commented:
"CRAW (compressed RAW)" 
craw does not mean compressed raw, but "compact raw"
raw (lossless) and craw (lossy) are both compressed

cytrinox committed:
Add credit for CR3 file structure information from lclevy/canon_cr3

cytrinox commented:
Added credit for CR3 file structure info from lclevy/canon_cr3, no other code changes are made.

cytrinox committed:
Improve black area detection and properly parse IAD1 box

IAD1 contains 4 definitions for rectangles:
 * crop rectangle
 * left optical black area rectangle
 * top optical black area rectangle
 * active area rectangle

This patch parses all IAD1 rectangles and uses them for
black areas and crop.

cytrinox commented:
I've added two more commits for fixing ISOSPEEDRATINGS integer overflow and better IAD1 box parsing.

The IAD1 box contains 4 rectangles, one for crop, two for black areas and one for active area. During tests a user reported that for the EOS RP model, black levels (from rawspeed calculation) are wrong for photos using larger ISO values (>= 800). For such images, the black areas include some white pixels which results in wrong black levels.

After some investigation, I think this is a camera bug, as the IAD1 box says blacklevel areas start at offset 0 (for vertial and horizontal) but in fact, there are white pixels located at this offset. 
I've written a workaround by shifting the black area offset by 12 pixels. This works well on RP and other tested models. If requested, I can enable this workaround only for the EOS RP model but I think the shift is safe for all models.

![grafik](https://user-images.githubusercontent.com/509535/119045198-6df92700-b9bb-11eb-801d-c5a3d29d48da.png)

johnny-bit committed:
Cleanup C to C++ conversion conventions

BodoHobbit commented:
Hi, I really appreciate your work for enabling CR3 support on darktable (especially since my old Canon camera is broken and now I switched to an EOS R). Is there still a chance to get CR3 support in the upcoming darktable release or is it already too late? Is ther anything I can help with? BR, Bodo

johnny-bit commented:
the 3.6 release is already past the feature freeze. You can rely on experimental builds and tests if you're ok with that.

cytrinox committed:
Add support for missing Canon CR3 cameras

 * Canon EOS 250D
 * Canon EOS M6 Mark II
 * Canon EOS M200
 * Canon EOS M50 Mark II
 * Canon PowerShot G5 X Mark II
 * Canon PowerShot G7 X Mark III

cytrinox committed:
FIX: ISOSPEEDRATINGS is of type SHORT, larger ISO values stored in RECOMMENDEDEXPOSUREINDEX

cytrinox commented:
Rebased to current develop branch and added all missing cameras. The list of cameras is now complete. Each CR3 camera is in the list.
For testers: You need this patch, too: https://github.com/darktable-org/darktable/pull/9641

My work on this PR is now finished. For those are not able (or willing) to build experimental code, you can have a look at my other project: https://github.com/dnglab/dnglab
It's a FOSS DNG converter for Linux with support for CR3 files. The CR3/Crx decoder was rewritten and documented, but only RAW (not CRAW) files are supported.
Edit: RAW, CRAW, PQ-HDR and roll files are supported now.

pitbuster committed:
Address static analisys errors.

silb commented:
For my PowerShot G5 X Mark II firmware version 1.1.0 I had to make the following modification to this PR to be able to view the CR3 images in raw and craw mode:

```
diff --git a/src/librawspeed/decoders/Cr3Decoder.cpp b/src/librawspeed/decoders/Cr3Decoder.cpp
index 402debf8..6a1f6ba1 100644
--- a/src/librawspeed/decoders/Cr3Decoder.cpp
+++ b/src/librawspeed/decoders/Cr3Decoder.cpp
@@ -577,6 +577,8 @@ bool Cr3Decoder::isCodecSupported(const std::string& compressorVersion) const {
                                               // (raw)
          || compressorVersion ==
                 "CanonCR3_002/00.10.00/00.00.00" // CR3 of 1DX Mark III (craw)
+         || compressorVersion ==
+                "CanonCR3_001/01.09.00/00.00.00" // PowerShot G5 X Mark II
          || compressorVersion ==
                 "CanonCR3_001/00.09.00/00.00.00"; // EOS R, EOS RP, M50, 250D,
                                                   // 90D, M6 Mark II, M200
```

Edit: spellings

LibRaw commented:
> For my PowerShot G5 X Mark II firmware version 1.1.0 I had to make the following modification to this PR to be able to view the CR3 images in raw and craw mode:


@silb , could you please share some sample file(s)?

silb commented:
@LibRaw you can download the ISO samples here: <http://buggy.secnod.net/~stigi/canon_powershot_g5_x_mark_ii_raws.zip>

It is the same set of images I supplied to the [extended CR3 samples](https://discuss.pixls.us/t/call-for-extended-cr3-samples-all-iso-combinations-needed/24959/21?u=silb).

Edit: typo

LibRaw commented:
@silb thank you for the samples.  LibRaw (library) is not affected by the issue (because of less strict rules used when selecting CR3 track)

cytrinox commented:
I've added more strict checks to detect changes in the crx codec very early.

cytrinox commented:
@silb Will update the PR soon. Can you have a look at https://discuss.pixls.us/t/call-for-samples-of-canons-raw-burst-shots/26219/1 please? Your G5 cam supports burst film rolls and I need a set of samples. Not for all ISOs but a set of CRAW and RAW with ISO 100 and something high like ISO 6400 (or maximum possible) would be great.

silb commented:
@cytrinox see <https://discuss.pixls.us/t/call-for-samples-of-canons-raw-burst-shots/26219/2?u=silb> for raw burst mode samples.

Thank you for your work on CR3 support.

paolobenve commented:
This PR yet shows no previews for CR3 files in import dialog, as I reported in https://github.com/darktable-org/darktable/issues/9935

cytrinox commented:
Previews are not related to rawspeed and/or this PR. The preview in the import dialog in darktable uses (if possible) the thumbnail embedded in the RAW file and this is handled by exiv2 only. I recommend to check the exiv2 code if thumbnail extraction for CR3 is already implemented but afaik it's not.

lclevy commented:
> Previews are not related to rawspeed and/or this PR. The preview in the import dialog in darktable uses (if possible) the thumbnail embedded in the RAW file and this is handled by exiv2 only. I recommend to check the exiv2 code if thumbnail extraction for CR3 is already implemented but afaik it's not.

This script can extract cr3 thumbnails
https://github.com/lclevy/canon_cr3/blob/9a2a9639ae7188d4dd70c1251df35f5af8ebee3d/parse_cr3.py

paolobenve commented:
> This script can extract cr3 thumbnails https://github.com/lclevy/canon_cr3/parse_cr3.py

It doesn't exist. See https://github.com/lclevy/canon_cr3/blob/master/parse_cr3.py instead.

Actually:

```
$ exiv2 -ep1 /path/to/my/image.cr3
/path/to/my/image.cr3: Image does not have preview 1

```

piratenpanda commented:
For everyone building with latest darktable git, cytronix' canon_cr3 branch needs those changes: https://github.com/darktable-org/rawspeed/commit/8ca1dfa7acc098b0ff17a314fff59762c68071fd and the respective changes in  src/librawspeed/decoders/Cr3Decoder.cpp,  src/librawspeed/decoders/Cr3Decoder.h, IsoMParser.cpp and most likely some files I forgot but the compiler will tell you where to replace "*"s with "&"s

cytrinox committed:
Use Buffer only as reference, not as pointer

cytrinox committed:
FIX: Building issue with darktable

cytrinox committed:
CR3: Add missing compressor versions that are actually supported

paolobenve commented:
what prevents this PR from being merged?

dhoulder commented:
https://github.com/Exiv2/exiv2/commit/b385f2db1cf8b11c7c0513d3140167d8ca98eca6 should fix the missing previews problem. It should appear in the next stable release of exiv2

mfg92 commented:
@LebedevRI I would be interested in knowing if it would be possible to merge this PR before the darktable 3.8 freeze.

I have been productively using this patch for over 5 month. From a users perspective it works fine and I have not found any issues with it.
Needless to say that I am not the only one who is using it, when looking over at https://github.com/darktable-org/darktable/issues/2170 one can see that there are many others who have tried it. Also over at _pixls.us_ are three threads ([1](https://discuss.pixls.us/t/current-win10-build-incl-cr3-support/25212), [2](https://discuss.pixls.us/t/darktable-3-6-0-unofficial-build-with-cr3-support-osx-win10/25703), [3](https://discuss.pixls.us/t/darktable-3-6-1-unofficial-build-with-cr3-support-osx-win10/26970)) providing builds of darktable with CR3 support. In total they have close to 5k views. 

So to me it looks like this code is already tested by many and works fine. So i am currious to know why this PR is not merged yet.

PS: Thank you for your great work!

paolobenve commented:
> So to me it looks like this code is already tested by many and works fine.

to me too: I have been using it for 6 months without any issue

axxel commented:
> what prevents this PR from being merged?

Since quite a few people seem to be asking themself this question and the single maintainer of this project has not commented on this PR, yet, I'd like to point to two comments buried in the history of this project. One giving an insight into said maintainers general priorities regarding this project (https://github.com/darktable-org/rawspeed/commit/78307bcb8726a897631ce13d599365d15282e817#commitcomment-20894267). The other providing some background on how he and the original author of the core code in this specific PR have parted ways (https://github.com/darktable-org/rawspeed/issues/121#issuecomment-460830865).

Disclaimer: I apologize in advance for this cheap attempt of luring him into "proving me wrong" by merging this PR or at least giving it some substantiated and well deserved feedback. I still hope it works. ;) And as for the tone, the first of the above links proves that I'm totally biased here.

Djip007 commented:
In fact try to build this PR after rebase on last develop branch... with last change for modern C++11/14 it do not build without changes...
I think I success correct it... but not to much test...

Change I need to build are:
```
# change all 
-  DEBUG_PRIO_ZZZZZ,
+ DEBUG_PRIO::ZZZZZ,

# and things like that (Add TiffTag:: )
-  if (canonBox->CMT2()->mRootIFD0->hasEntryRecursive(ISOSPEEDRATINGS)) {
+  if (canonBox->CMT2()->mRootIFD0->hasEntryRecursive(TiffTag::ISOSPEEDRATINGS)) {

# last update:
-  if (mRaw->getDataType() != TYPE_USHORT16)
+  if (mRaw->getDataType() != RawImageType::UINT16)
```

Need more time to test and figure if/how I can help with that.

nm17 commented:
Waiting for an update on this

piratenpanda commented:
Cr3 Files will be supported in 3.8 via libraw

cytrinox committed:
rawspeed: Canon EOS 90D support

cytrinox commented:
@LebedevRI Rebased and added ColorMatrix tag for cameras.

cytrinox committed:
Fix compiler errors for unused check() result

Cleanup reserved macro and function names for llvm

cytrinox committed:
Restore RangeTest.h from develop branch

cytrinox committed:
Make code compatible to latest develop branch

cytrinox committed:
Add ColorMatrix tag for CR3 cameras

axxel commented:
Glad to see this effort has not died with the libraw path for CR3 in Darktable now!

CameronMcGehee commented:
Hope this gets implemented. This would be incredibly useful!

paolobenve commented:
darktable _already_ opens CR3 files.

johnny-bit commented:
BUT afaik not via rawspeed yet.

pt., 24 cze 2022 o 22:59 Paolo Benvenuto ***@***.***>
napisał(a):

> darktable *already* opens CR3 files.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/darktable-org/rawspeed/pull/271#issuecomment-1165934059>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACRRKFIUSQQ5YUPW2G2DTEDVQYOTJANCNFSM44WZTUVA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
Pozdrawiam,
Hubert Kowalski

CameronMcGehee commented:
> darktable _already_ opens CR3 files.

I have tried and it is not showing up in the list of supported formats, and throws an error when I force it to open a CR3 file...my system is up to date...am I doing something wrong?

matt-maguire commented:
> I have tried and it is not showing up in the list of supported formats, and throws an error when I force it to open a CR3 file...my system is up to date...am I doing something wrong?

It depends which distro you are using and how you installed darktable -- by default, CR3 support is disabled in the EXIV2 library, and the person packaging exiv2/darktable for your distribution needs to take explicit steps to enable the CR3 support.

CameronMcGehee commented:
> > I have tried and it is not showing up in the list of supported formats, and throws an error when I force it to open a CR3 file...my system is up to date...am I doing something wrong?
> 
> It depends which distro you are using and how you installed darktable -- by default, CR3 support is disabled in the EXIV2 library, and the person packaging exiv2/darktable for your distribution needs to take explicit steps to enable the CR3 support.

Oh ok. It seems as though my darktable, even when installed fresh, is still on version 3.0.1. I tried installing the latest version from the .deb file on the website but it says a dependency (libc6) can't be satisfied. I tried installing libc6 and it won't install the latest version for some reason. Why are all these package managers not updated by the devs when new versions come out? I thought that was the whole point of a package manager...

TheBB commented:
Are you using Ubuntu 20.04 or some equivalent OS?

For a fixed version of Ubuntu, new versions of packages are normally only pushed for bugfixes and security updates. If you want newer software, either use a newer version of Ubuntu (22.04 perhaps) or get packages via other means better for bleeding edge, like Flathub, PPAs or building them yourself.

CameronMcGehee commented:
> Are you using Ubuntu 20.04 or some equivalent OS?
> 
> For a fixed version of Ubuntu, new versions of packages are normally only pushed for bugfixes and security updates. If you want newer software, either use a newer version of Ubuntu (22.04 perhaps) or get packages via other means better for bleeding edge, like Flathub, PPAs or building them yourself.

Hmm OK thanks, I see I am still on 20.04. Why doesn't it update to 22.04 from the update command? How does one update without doing a fresh install? Is this even possible?

Edit: I see now it's because Linux Mint is in control of what version of Ubuntu is running...and they haven't yet updated. I guess that explains why people usually get more and more towards the source as they get more familiar with Linux...

mallman commented:
Hi all. It looks like this PR has gone a little stale. Is this still being planned for merge?

paolobenve commented:
CR3 support is inside darktable since some version

mallman commented:
> CR3 support is inside darktable since some version

I'm aware of that. It uses LibRaw for that. I'm referring to support for CR3 images in RawSpeed, i.e. this PR.

hanatos commented:
i have in the past rebased this cr3 branch onto newer versions of the develop branch. lately this has become harder than usual, so i had to actually look at the code in doing so.

@LebedevRI would you be interested in merging this feature, given some overhaul and cleanup? if so, i could try to spend some time on this. i don't have a cr3 camera myself but i think this is a major feature and i think rawspeed is the most sophisticated and well engineered library available and should be able to load cr3.

i was able to open a cr3 with this rebase of the cytrinox/canon_cr3 branch today: https://github.com/hanatos/rawspeed/tree/rebase_cr3

but i would have to say i did a few fairly terrible things to make it work. i would completely understand if you reject to merge this (to a point that i don't think it's ready for an official pull request). but if there is general interest i'll be happy to take suggestions how to improve this.

some thoughts for discussion:
- need to resolve pointer alignment issues
- can we reuse a bitstream from rawspeed's code base instead of the one implemented in CrxDecompressor here?
- what about c-style allocation schemes here?
- probably a lot about fuzzing/testing/coverage etc that i don't understand and would need help with
- do you prefer a full rewrite (from fresh libraw code) even over a temporary solution?

seanpkent commented:
Any possibility of this PR being merged in the near future?

I am also using a fork of Rawspeed with this CR3 support merged in. Rawspeed is an amazing and fast raw decoder. I would love to have CR3 baked in by default. Any contributions I can make to see this happen?

axxel commented:
For those who are recently asking themselves why this is not moving anywhere: I made this [comment above](https://github.com/darktable-org/rawspeed/pull/271#issuecomment-966997588) 2 years ago, hinting at some speculative theory as to why the sole maintainer of this repo has not provided a single comment in this 2,5 year old discussion here.

I am personally not yet depending on CR3 support and am still running a fork from around 6 years ago, when I left the project. This might change soon. As things are looking right now, I personally have zero hope that Roman will merge this. Maybe there would be interest from a small group of interested parties to focus on a 'community' fork with CR3 support?

dhoulder commented:
To play devil's advocate for a moment, Roman did actually open https://github.com/darktable-org/rawspeed/issues/121 and moved it  to "Code needs porting" in "New Raw Formats"  in December last year, so I doubt he's opposed to the idea. I'm also glad to see him putting work into fuzzing, CI and code cleanup. Having said that, I agree that he's being overly taciturn in regard to this PR. We also have a viable alternative for the time being in the form of libraw, which handles CR3 files just fine in my experience.

However, I also think that the lack of CR3 support doesn't bode well for the future of rawspeed as a project. With Canon still having a dominant position in the market, I'd bet that CR3 would now be one of the most common raw formats. In this respect, rawspeed is definitely not meeting its goal of "support[ing] the most common DSLR and similar class brands." (https://github.com/darktable-org/rawspeed/blob/develop/README.rst )

darktable will almost certainly continue to use libraw for CR3 support for the foreseeable future, and at some point a developer will probably ask "why are we using two raw decoding libraries when libraw could do it all", followed by good arguments about economies of scale and not reinventing wheels.

Rawspeed might be faster and better engineered than libraw, but if it can't do what libraw does, it will lose out. It would not be the first time in history that the more elegant technology lost out to the thing that was "good enough".

> Maybe there would be interest from a small group of interested parties to focus on a 'community' fork with CR3 support?

Possibly. I worked on the CR3 code in exiv2, so I could be interested. I think that would be a fairly major step though. A lot of ongoing work would be required.

kmilos commented:
> darktable will almost certainly continue to use libraw for CR3 support for the foreseeable future, and at some point a developer will probably ask "why are we using two raw decoding libraries when libraw could do it all", followed by good arguments about economies of scale and not reinventing wheels.

This is unlikely to happen either, because of LibRaw's lack of roadmap transparency/commitment and (equally problematic) outside developer inclusion - will happily take patches, but only release public _betas_ w/ new camera support once a year (maybe?), and stable releases only every two years or so (which is not great for distribution, as most would prefer to use the stable system package rather than a vendored beta).

So, dt is simply stuck w/ both of these imperfect solutions for that foreseeable future unfortunately.

FYI, [vkdt seems to be leaning towards OP's new rawler library](https://discuss.pixls.us/t/released-vkdt-0-7-0/40017).

axxel commented:
Question: what was the reason to add LibRaw next to the existing RawSpeed in the first place?

kmilos commented:
> Question: what was the reason to add LibRaw next to the existing RawSpeed in the first place?

It was re-enabled two years ago just because of CR3.

Originally (it was added years ago and then disabled for a very long time), I don't know why.

axxel commented:
> It was re-enabled two years ago just because of CR3.

I see. So do I get this right?
 * darktable-org is comprised of a whole bunch of people and is maintaining the darktable and the rawspeed repository
 * darktable was in need for a CR3 decoder
 * there was a working PR available for RawSpeed at the time, waiting for review
 * because Roman did simply ignore it, someone else came to the conclusion that the best cause of action is to bring LibRaw back into the mix

Why did then not someone else step in and merge this PR? If LibRaw's only purpose in dt is to provide CR3 support and this situation is viewed as "unfortunate" why then not simply move ahead with this PR, where multiple people offered to help with open issues/questions, and end up with a working CR3 decoder in RawSpeed?

kmilos commented:
> darktable-org is comprised of a whole bunch of people and is maintaining the darktable and the rawspeed repository

> Why did then not someone else step in and merge this PR?

I think this was already covered. The darktable and rawspeed repos are not maintained the same way.

axxel commented:
> I think this was already covered.

I don't know what you are referring to here, sorry if I missed something.

> The darktable and rawspeed repos are not maintained the same way.

Thanks for answering. So I understand that I was wrong in at least in one part: even though the rawspeed repo is owned by darktable-org, this does not mean that any of the 27 people in that group other than Roman have the 'physical' ability or 'organizational' role to make changes to that repo?

If the dt stakeholders and the rawspeed stakeholder are effectively two "separate" groups with separate interests, I wonder if the dt people would also potentially be interested in a 'community' fork that contains CR3 support?

kmilos commented:
> > I think this was already covered.
> 
> I don't know what you are referring to here, sorry if I missed something.

Just to the same conclusion you have just reached. Again. 😉 (or 😞)

mallman commented:
I have extensive experience working with both Libraw and Rawspeed. For one thing, contrary to some belief, Libraw support for raw files is not a superset of Rawspeed. At least, I have some raw files which Rawspeed can decode but which Libraw cannot.

Second, I have published a sort of "compatibility interface" for raw file decoding which provides a common API to decode raw files with Libraw and/or Rawspeed. I wrote it for an iOS app, so it is written in Objective C. That makes it suitable for the Apple universe of projects, which is intentional. Anyway, this project is public and available at https://github.com/mallman/RawDecoder. If you're looking for a C or C++ interface, this is not the place for you. :(

Third, I prefer RawSpeed over Libraw for many of the same reasons cited here. And I see *many* opportunities for improvement and optimization—and I would be happy to contribute code. However, I don't expect this project to accept any outside contribution, so I haven't bothered contributing anything. Additionally, the license is too encumbered. Unless someone can magically fork this into a repo that doesn't use a copyleft license, I will not be a part of the fun.

Personally, I would prefer a community driven open source codebase without a copyleft license. That is not Rawspeed. I looked for something like this for a long time, and didn't find anything. I will have a look at vkdt and rawler. I don't think I've seen them before. Thanks for the tip.

LibRaw commented:
> I have extensive experience working with both Libraw and Rawspeed. For one thing, contrary to some belief, Libraw support for raw files is not a superset of Rawspeed. At least, I have some raw files which Rawspeed can decode but which Libraw cannot.
> 

Could you please provide problematic sample files  via  LibRaw issues on GitHub ( https://github.com/LibRaw/LibRaw/issues ) or via E-mail to info@libraw.org?

Thanks in advance.

kmilos commented:
> Personally, I would prefer a community driven open source codebase without a copyleft license. That is not Rawspeed. I looked for something like this for a long time, and didn't find anything. I will have a look at vkdt and rawler. I don't think I've seen them before. Thanks for the tip.

There's also [libopenraw](https://libopenraw.freedesktop.org/), but I guess doesn't fit your licensing requirements either.

jcampbell05 commented:
If progress has stalled on this, does anyone know if we could build a framework under the Linux Foundation or some kind of foundation like that

MStraeten commented:
> we could build a framework under the Linux Foundation or some kind of foundation like that

Who is that ‚we‘ that takes the responsibility maintaining that for a whole bunch of platforms? At least no one stepped in and forked a rawspeed-nextgen project. And also no one stepped in to replace libraw within darktable with rawler which is a bit faster in supporting new canon cameras …

jcampbell05 commented:
I can find this library you speak of but it's all the same problem, we could opt for super camera deluxe library. But if there isn't a clear structure around making sure it's supported and well funded then it will fizzle out eventually even if it's pretty good now. 

Libraw has lasted as long as it has because it has a well defined funding model, even if that funding model hurts the time it takes to adopt new standards

Ideally the community could find a solution for that.

MStraeten commented:
there’s no developer or maintainer named ‘the community’ in the rawspeed project ;)

jheidemann commented:
I added a patch for Canon EOS M6 Mark II at #284 

I'm commenting here as that ticket recommends.  Can someone with commit access please merge the patch?
(Thanks in advance.)

axxel commented:
Is anyone here aware of a rawspeed fork that has a more up-to-date version of this PR / CR3 support, especially including support for EOS R8 and EOS R5 Mark II?

MStraeten commented:
if you need that for darktable you need to take a look at libraw requiring a custom build following  https://blog.pandainthecloud.de/2024/09/07/adding-new-camera-support-in-darktable-libraw-exiftool/ 
\- except you're on macos, then you can use https://discuss.pixls.us/t/current-osx-build/13213/840

axxel commented:
> if you need that for darktable you need to take a look at libraw requiring a custom build following

No, I am actually looking for rawspeed code (like this PR contains) but wondering whether someone worked on it in the meantime in a fork (since this PR here is apparently going nowhere). But thanks a lot for that link anyway. Might come in handy... ;)

EDIT: for whom it may concern... I backported this PR into my old fork and the only thing I had to change to make it work for the R8 was [this line](https://github.com/axxel/rawspeed/blob/e6e388e427ba4484a24257cf0ffb8899e02d798d/src/librawspeed/decoders/Cr3Decoder.cpp#L581). I have not looked at the R5m2, yet.

mcourteaux commented:
I revived this PR personally and updated everything to the latest RawSpeed version. My conclusions are the following:

 - This is a port from libraw, which has a __very__ different code style. This code is almost unreadable and unmaintainable in the context of RawSpeed.
 - It has bugs which I didn't try to track down. I managed to load a couple of test files, but others caused crashes in the IsoMBox Tiff parsing.
 - I don't see the benefit of porting over the code from @alextutubalin to RawSpeed, and then not being able to maintain it (as the resulting port is still drastically different and nothing is copy-pastable).

LibRaw commented:
> * I managed to load a couple of test files, but others caused crashes.

Could you please share 'crashing' files to test with LibRaw code?

If there is a problem with the original decoder (which is very unlikely, since it is used by hundreds of thousands of users), then we will of course improve it.

mcourteaux commented:
> If there is a problem with the original decoder (which is very unlikely, since it is used by hundreds of thousands of users), then we will of course improve it.

The crashes I was seeing were happening in the IsoMBox Tiff decoder stuff. Not anything related to the actual decompressor. I'll integrate LibRaw soon, and then test again. I expect no issues with LibRaw itself. Thanks for your quick response @LibRaw.

LibRaw commented:
> IsoMBox Tiff decoder stuff.

Is this code ported from LibRaw?

axxel commented:
> I revived this PR personally and updated everything to the latest RawSpeed version. My conclusions are the following:

Interesting read. Thanks for sharing. Is [this](https://github.com/mcourteaux/rawspeed/commits/canon_cr3/) the code you've been talking about here?

>     * It has bugs which I didn't try to track down. I managed to load a couple of test files, but others caused crashes in the IsoMBox Tiff parsing.

The IsoMBox code is mostly from Roman with some change from @cytrinox and has nothing to do with the LibRaw code (if I'm not mistaken). I found the readability of that code to be lacking and would personally replace it by something simple that feels less "over engineered". I guess any crash caused by it should be fixable relatively easy, though.

>     * I don't see the benefit of porting over the code from @alextutubalin to RawSpeed, and then not being able to maintain it (as the resulting port is still drastically different and nothing is copy-pastable).

When I was trying to find out if/what had changed in the LibRaw code between this PR and today (looking for R8 support), I came to the opinion that the best approach to integrate this code would be to take it as is in LibRaw and modify it as little as possible by supplying some type of wrapper code to map between the different memory management APIs and thereby being able to basically share the same code between the two projects. From the perspective of a rawspeed user this should be preferable compared to having to wrap the complete rawspeed and LibRaw libraries just to get access to CR3 files. This would be on the premise that LibRaw will continue maintaining this code and there is no reason to fundamentally change it (like finding out that it can be made substantially faster for example).

LibRaw commented:
> take it as is in LibRaw and modify it as little as possible

And it should be very easy to do, since CR3 decoder itself is completely separate from the rest of the LibRaw library.
You only need to replace the data reading and feed it the header from the desired track for parsing (plus output buffer allocated externally).

piratenpanda commented:
Any chance we will see cr3 support in rawspeed?
